pm/recipient_test.go

```diff
@@ -0,0 +1,436 @@
+package pm
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"math/big"
+	"testing"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+func TestReceiveTicket(t *testing.T) {
+	sender := ethcommon.HexToAddress("A69cdA26600c155cF2c150964Bdb5371ac3f606F")
+	b := newStubBroker()
+	v := &stubValidator{}
+	ts := newStubTicketStore()
+	secret := [32]byte{3}
+	faceValue := big.NewInt(100)
+	winProb := big.NewInt(100)
+	sig := []byte("foo")
+
+	r := NewRecipient(b, v, ts, secret, faceValue, winProb)
+
+	// Config stub validator with valid non-winning tickets
+	v.SetIsValidTicket(true)
+
+	params, err := r.TicketParams(sender)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Test invalid recipientRand from seed (invalid seed)
+	ticket := &Ticket{
+		Recipient:         ethcommon.Address{},
+		Sender:            sender,
+		FaceValue:         faceValue,
+		WinProb:           winProb,
+		SenderNonce:       0,
+		RecipientRandHash: params.RecipientRandHash,
+	}
+
+	// Using invalid seed
+	_, err = r.ReceiveTicket(ticket, sig, big.NewInt(12312))
+	if err == nil {
+		t.Error("expected invalid recipientRand from seed error")
+	}
+	if err != nil && err != errInvalidRecipientRandFromSeed {
+		t.Errorf("expected invalid recipientRand from seed error, got %v", err)
+	}
+
+	// Test invalid recipientRand from seed (invalid sender)
+	ticket = &Ticket{
+		Recipient:         ethcommon.Address{},
+		Sender:            ethcommon.Address{}, // Using invalid sender
```

Comment:
This is confusing - can you help me understand how adding an invalid sender tests for the use case of an invalid recipientRand?

I thought testing for an invalid recipientRand would look more like providing an invalid value for `RecipientRandHash`. Perhaps the invalid sender scenario should have a test of its own with its own error?

Reply:
While the seed is valid since it was received from the recipient, the sender address included in the ticket is invalid such that a invalid `recipientRand` is generated since `HMAC(recipient.Secret, seed | T.Sender) != T.RecipientRandHash`.

I could add an additional test where `T.RecipientRandHash` is some nonsense value as opposed to the value included in the `TicketParams` returned by a recipient.

Reply:
Ok, that makes sense to me now! I think the additional test you suggested could be nice!